### PR TITLE
Fix issue where the status code set by pf::WebAPI::JSONRPC did not match

### LINKED ISSUE #929: JSONRPC client is not reading the response code properly
Server returns 500 in the httpd.webservices.access
JSONRPC client sees a 200

As Zammit said, most likely due to a recent change....

This breaks the cluster sync error detection

### PR DESCRIPTION
**Pull request by jrouzierinverse**
Fixes #929
## Description

If an exception happened when handling the rpc request the error sent by apache did match the one set by pf::WebAPI::JSONRPC
## Impacts

The pf::api::jsonrpcclient response handlin
## NEWS file entries

Bug Fixes
+++++++++
- Fix improper httpd status code being set
## Delete branch after merge

NO

### CODE REVIEW

File: lib/pf/WebAPI/JSONRPC.pm [hunk 1]
```diff
@@ -15,99 +16,142 @@ use strict;
...
 use JSON::XS;
 use pf::log;
-use Apache2::RequestIO();
-use Apache2::RequestRec();
-use Apache2::Response ();
-use Apache2::Const -compile => qw(OK DECLINED HTTP_UNAUTHORIZED HTTP_NOT_IMPLEMENTED HTTP_UNSUPPORTED_MEDIA_TYPE HTTP_PRECONDITION_FAILED HTTP_NO_CONTENT HTTP_NOT_FOUND SERVER_ERROR HTTP_OK HTTP_INTERNAL_SERVER_ERROR);
+use Apache2::RequestIO;
+use Apache2::RequestRec;
+use Apache2::Response;
+use Apache2::Log;
```
Comment (fdurand) [inline lib/pf/WebAPI/JSONRPC.pm]: Useless since you are not using apache log ($r->log)

**Comment by jrouzierinverse** [comment]
Review me

File: lib/pf/WebAPI/JSONRPC.pm [hunk 2]
```diff
@@ -15,99 +16,141 @@ use strict;
...
-    my $ref_type = ref($data);
-    return Apache2::Const::HTTP_UNSUPPORTED_MEDIA_TYPE unless $ref_type;
-    #for now not supporting batch
-    return Apache2::Const::HTTP_NOT_IMPLEMENTED if $ref_type eq 'ARRAY';
-    my ($method,$id,$jsonrpc) = @{$data}{qw(method id jsonrpc)};
-    return Apache2::Const::HTTP_NOT_IMPLEMENTED  unless defined $method;
-    my ($response, $status_code, $method_sub, @args);
+    my ($self, $r) = @_;
+    my $content = $self->get_all_the_content($r);
+    my $data = eval {decode_json $content };
```
Comment (julsemaan) [inline lib/pf/WebAPI/JSONRPC.pm:49]: adjust the whitespace inside the brackets

File: lib/pf/WebAPI/JSONRPC.pm [hunk 3]
```diff
@@ -15,99 +16,141 @@ use strict;
...
+        return $self->send_request(
+            $r,
+            Apache2::Const::HTTP_UNSUPPORTED_MEDIA_TYPE,
+            $self->make_error_object(undef, $JSONRPC_ERROR_CODE_INVALID_REQUEST, "Invalid request\n"),
+        );
+    }
+
+    my ($method, $id, $jsonrpc) = @{$data}{qw(method id jsonrpc)};
+    unless (defined $method) {
+        get_logger->error("Invalid request no method defined");
```
Comment (julsemaan) [inline lib/pf/WebAPI/JSONRPC.pm]: Invalid request, no...

File: lib/pf/WebAPI/JSONRPC.pm [hunk 4]
```diff
@@ -15,99 +16,141 @@ use strict;
...
+        PerlCleanupHandler => sub {
+            eval {$dispatch_to->$method(@args);};
+            $logger->error($@) if $@;
+        }
+    );
+    $r->content_type($CONTENT_TYPE);
+    return Apache2::Const::HTTP_NO_CONTENT;
+}
+
+sub send_request {
```
Comment (julsemaan) [inline lib/pf/WebAPI/JSONRPC.pm]: This should be named send_response not request

File: lib/pf/WebAPI/JSONRPC.pm [hunk 5]
```diff
@@ -15,99 +16,141 @@ use strict;
...
+    $r->custom_response($status, '');
+    $r->content_type($CONTENT_TYPE);
+    $r->status($status);
+    my $response_content = encode_json($object);
+    $r->print($response_content);
+    $r->rflush;
     return Apache2::Const::OK;
 }
 
+sub get_all_the_content {
```
Comment (julsemaan) [inline lib/pf/WebAPI/JSONRPC.pm]: get_all_content ?

File: lib/pf/api.pm [hunk 6]
```diff
@@ -933,6 +933,17 @@ sub fingerbank_submit_unmatched : Public {
     pf::config::util::pfmailer(( subject => 'Fingerbank - Submit unknown/unmatched fingerprints status', message => $status_msg ));
 }
 
+=head2 throw
+
+TODO: documention
```
Comment (julsemaan) [inline lib/pf/api.pm]: do it :)

File: lib/pf/WebAPI/JSONRPC.pm (hunk 3 above)
Comment (rouzier) [inline lib/pf/WebAPI/JSONRPC.pm]: 'Invalid request no method defined'

Comment (julsemaan) [inline lib/pf/WebAPI/JSONRPC.pm]: Missing a comma I think.. That was my point

Comment (jrouzierinverse) [inline lib/pf/WebAPI/JSONRPC.pm]: It was unclear what you wanted

**Comment by jrouzierinverse** [comment]
Please review

**Pull request merged**